bleak/backends/winrt/client.py

```diff
@@ -0,0 +1,835 @@
+                                    descriptors_result.protocol_error,
+                                    CONTROLLER_ERROR_CODES.get(
+                                        descriptors_result.protocol_error, "Unknown"
+                                    ),
+                                )
+                            )
+                        else:
+                            raise BleakDotNetTaskError(
+                                "Could not get GATT descriptors for {0}: {1}".format(
+                                    characteristic,
+                                    _communication_statues.get(
+                                        descriptors_result.status, ""
+                                    ),
+                                )
+                            )
+                    for descriptor in list(descriptors_result.descriptors):
+                        self.services.add_descriptor(
+                            BleakGATTDescriptorWinRT(
+                                descriptor,
+                                str(characteristic.uuid),
+                                characteristic.attribute_handle,
+                            )
+                        )
+
+            logger.info("Services resolved for %s", str(self))
+            self._services_resolved = True
+            return self.services
+
+    # I/O methods
+
+    async def read_gatt_char(
+        self,
+        char_specifier: Union[BleakGATTCharacteristic, int, str, uuid.UUID],
+        use_cached=False,
+        **kwargs
+    ) -> bytearray:
+        """Perform read operation on the specified GATT characteristic.
+
+        Args:
+            char_specifier (BleakGATTCharacteristic, int, str or UUID): The characteristic to read from,
+                specified by either integer handle, UUID or directly by the
+                BleakGATTCharacteristic object representing it.
+            use_cached (bool): ``False`` forces Windows to read the value from the
+                device again and not use its own cached value. Defaults to ``False``.
+
+        Returns:
+            (bytearray) The read data.
+
+        """
+        if not isinstance(char_specifier, BleakGATTCharacteristic):
+            characteristic = self.services.get_characteristic(char_specifier)
+        else:
+            characteristic = char_specifier
+        if not characteristic:
+            raise BleakError("Characteristic {0} was not found!".format(char_specifier))
+
+        read_result = await characteristic.obj.read_value_async(
+            BluetoothCacheMode.CACHED if use_cached else BluetoothCacheMode.UNCACHED
+        )
+
+        if read_result.status == GattCommunicationStatus.SUCCESS:
+            reader = DataReader.from_buffer(read_result.value)
+            # TODO: Figure out how to use read_bytes instead...
+            value = bytearray(
+                [reader.read_byte() for _ in range(reader.unconsumed_buffer_length)]
```

Comment:
I made an interesting find. No reader required.

```python
from winrt.windows.security.cryptography import CryptographicBuffer
...
value = bytearray(CryptographicBuffer.copy_to_byte_array(read_result.value))
```

The `copy_to_byte_array` method returns a list of integers.

Reply:
A `CryptographicBuffer`?! Why yes, it makes sense, but I would never have though of looking for what I needed there...

It does introduce an element of surprise to the potential code reader though. I would have wondered why the developers had chosen to do this when it had nothing to do with cryptography.

It is almost three times as fast as well. Good find!

Reply:
I came across this class before because it is the only place in .NET where there is a byte array to hex string conversion function.